### ISSUE TITLE
OOMs on Idle Cluster running rc3

### ISSUE THREAD

**Issue by tim-o** [cockroachdb/cockroach#42331]
Cluster running rc3, on 6x n1-standard-16 nodes is regularly OOMing on each node, despite the cluster being idle and not under load.

Notes on how to access the cluster and logs from a single node can be found here: https://cockroachdb.zendesk.com/agent/tickets/4057

We're working on collecting a debug zip and profiles as a next step, but it'll likely wait till Monday.

Some graphs below that seemed interesting - low CPU utilization, rapidly increasing memory utilization. No SQL queries or connections. No service latency recorded, but we do see regular multi-second spikes in execution latency. p99 KV transaction duration is multiple seconds, and we do see some lengthy queue processing times (specifically in the replication and raft log queues).

![Screen Shot 2019-11-08 at 6 12 38 PM](https://user-images.githubusercontent.com/38867162/68517244-c42ccc00-0254-11ea-82ae-482f7db51ec5.png)
![Screen Shot 2019-11-08 at 6 12 00 PM](https://user-images.githubusercontent.com/38867162/68517245-c42ccc00-0254-11ea-90d7-4f79ca01047a.png)
![Screen Shot 2019-11-08 at 6 09 09 PM](https://user-images.githubusercontent.com/38867162/68517246-c42ccc00-0254-11ea-8230-69d390b62625.png)
![Screen Shot 2019-11-08 at 6 08 57 PM](https://user-images.githubusercontent.com/38867162/68517247-c42ccc00-0254-11ea-960c-6863286fb8b8.png)
![Screen Shot 2019-11-08 at 6 08 43 PM](https://user-images.githubusercontent.com/38867162/68517248-c42ccc00-0254-11ea-8d59-bc3adb7dc57b.png)
![Screen Shot 2019-11-08 at 6 08 34 PM](https://user-images.githubusercontent.com/38867162/68517250-c42ccc00-0254-11ea-847c-72a80c87c29d.png)
![Screen Shot 2019-11-08 at 6 05 03 PM](https://user-images.githubusercontent.com/38867162/68517251-c42ccc00-0254-11ea-80f2-9c84cf8d2776.png)
![Screen Shot 2019-11-08 at 6 04 44 PM](https://user-images.githubusercontent.com/38867162/68517252-c42ccc00-0254-11ea-96a5-5befd254a53e.png)
![Screen Shot 2019-11-08 at 6 04 09 PM](https://user-images.githubusercontent.com/38867162/68517253-c4c56280-0254-11ea-93de-f4a8b76c3aa3.png)

**Comment by RoachietheSupportRoach** [comment]
Zendesk ticket #[4057](https://cockroachdb.zendesk.com/agent/tickets/4057) has been linked to this issue.

**Comment by nvanbenschoten** [comment]
> We're working on collecting a debug zip and profiles as a next step

Yep, heap profiles are exactly what we'll need to track this down.

**Comment by bdarnell** [comment]
@nvanbenschoten There are heap profiles and goroutine dumps (from a single node) in the logs tarball in zendesk (tim's message yesterday). The memory is all coming from scans; I haven't been able to deduce what's causing the scans or why the memory isn't getting freed. 

```
Showing top 10 nodes out of 59
      flat  flat%   sum%        cum   cum%
   10.58GB 90.73% 90.73%    10.58GB 90.73%  github.com/cockroachdb/cockroach/pkg/storage/engine.copyFromSliceVector
    0.94GB  8.04% 98.77%     0.94GB  8.04%  github.com/cockroachdb/cockroach/pkg/roachpb.(*ScanResponse).Unmarshal
    0.06GB  0.54% 99.31%     0.06GB  0.54%  github.com/cockroachdb/cockroach/vendor/github.com/andy-kimball/arenaskl.NewArena
         0     0% 99.31%     0.08GB  0.68%  github.com/cockroachdb/cockroach/pkg/cli.runStart.func3
         0     0% 99.31%     0.08GB  0.68%  github.com/cockroachdb/cockroach/pkg/cli.runStart.func3.2
         0     0% 99.31%     0.94GB  8.04%  github.com/cockroachdb/cockroach/pkg/internal/client.(*DB).sendUsingSender
         0     0% 99.31%    11.51GB 98.67%  github.com/cockroachdb/cockroach/pkg/kv.(*DistSender).Send
         0     0% 99.31%    11.51GB 98.67%  github.com/cockroachdb/cockroach/pkg/kv.(*DistSender).divideAndSendBatchToRanges
         0     0% 99.31%    11.51GB 98.67%  github.com/cockroachdb/cockroach/pkg/kv.(*DistSender).sendPartialBatch
         0     0% 99.31%    11.51GB 98.67%  github.com/cockroachdb/cockroach/pkg/kv.(*DistSender).sendRPC
```

**Comment by bdarnell** [comment]
![image](https://user-images.githubusercontent.com/160562/68534815-fdbe0f80-0306-11ea-8c61-07d29eabf1db.png)

The main thing that appears to be happening on the cluster is table stats refreshes, which are running for a time then getting cancelled (hard to tell whether these jobs are the root cause of the problems or just a symptom). The stats jobs are running on two tables, each of which contains a few hundred GB of data (and these two tables are the only ones in the cluster with a non-trivial amount of data)

**Comment by bdarnell** [comment]
I would suggest setting `sql.stats.automatic_collection.enabled` to false to see if that stabilizes the cluster.

**Comment by RaduBerinde** [comment]
CC @rytaft. We should investigate why the auto stats are failing.

**Comment by bdarnell** [comment]
My read of the profiles is that something is holding on to the original encoded bytes of a datum when they shouldn't be. This leads to an amplification in memory use since we're accounting for a single record but holding on to a whole block of memory. My best guess as to where this is happening is in the SampleReservoir (just because it's also doing some allocations). 

One possibility is that the assumptions in [`SampleReservoir.copyRow`](https://github.com/cockroachdb/cockroach/blob/9fd9e00e6c7912db09b52665af153c06537ebb98/pkg/sql/stats/row_sampling.go#L134) are not true any more. Perhaps an optimization to remove allocations or copies means that EnsureDecoded and DatumToEncDatum may not be enough to break all references to the original data. The types present in this table are INT8, STRING, and JSON (the JSON column is not indexed). 

Another possibility is that memory accounting for some data types (i.e. JSON) is imprecise and we're just using more memory than we're accounting for. I think it would need to be off by at least a factor of three, though, to account for the OOMs here.

**Comment by RaduBerinde** [comment]
Looking at the auto stats code, we don't try to get statistics for JSON columns. And we trim down sampled strings to 400bytes each.

**Comment by RaduBerinde** [comment]
Hm, we seem to be sampling the entire row of values, not just the columns for which we want histograms - this should definitely be improved. I don't think this explains anything here though.

Looking at the code in more detail, we should give up on histograms as soon as our accounted mem usage reaches `sql.distsql.temp_storage.workmem` (64MB). So if the accounting is off, it's off by a lot. I looked at the datum encoding/decoding and I can't find any way where it would hold on to a bigger slice.

**Comment by RaduBerinde** [comment]
This could be the problem in https://github.com/cockroachdb/cockroach/issues/33660. We're not doing anything wrong in the stats code, but we are requesting 10,000 keys of the table in each batch (just like a full table scan would). If the values are large, this could be a lot of unaccouned memory. The heap profile there also had `copyFromSliceVector` at the top.

**Comment by RaduBerinde** [comment]
That doesn't explain why this is not a problem in 19.1 and it is in 19.2 since nothing changed w.r.t. scan batch size. Except if we were already close to the edge and now we're using a bit more (due to histograms and whatnot).

**Comment by naveenwashere** [comment]
> I would suggest setting `sql.stats.automatic_collection.enabled` to false to see if that stabilizes the cluster.

I've made this change as of this comment.. Will check how things look on Monday.

**Comment by RaduBerinde** [comment]
Can we find out how many rows the two big tables have? (and how big they are in terms of Gb) Trying to ascertain the average row size.

**Comment by naveenwashere** [comment]
> Can we find out how many rows the two big tables have? (and how big they are in terms of Gb) Trying to ascertain the average row size.

We have two clusters, both running v19.2. In the cluster using 6 n1-standard-16 instances there's just one table that has 746.9 GiB data, in the form of 168891604 rows. This table has 3 indexes. Looks like after setting `sql.stats.automatic_collection.enabled` to `false` the pods seemed to have stopped restarting.

In the cluster using 6 n1-standard-8 instances there's two tables:
One has 350.1 GiB GiB data. This table has a few indexes.
The other has 109.7 GiB. This table has no indexes. 
The pods on this cluster still seem to be restarting every few mins/seconds, hence I'm unable to get the counts. But the table and the data is the same as the one I shared above. So the row size should be exactly the same.

**Comment by rytaft** [comment]
Thanks for this info, @naveenwashere -- doesn't sound like the rows are unusually large. And thanks @RaduBerinde and @bdarnell for the investigation so far. I'll look into this more... might be worth trying to only sample the histogram columns as @RaduBerinde suggested above since that should at least reduce the amount of memory overhead due to histograms. I'll try to make that change.

**Comment by rytaft** [comment]
We could be underestimating the memory allocation by a factor of 16 since the `DatumAlloc` allocates new datums in chunks of size 16.

**Comment by rytaft** [comment]
Seems unlikely that's the cause of the OOM, though, since all the heap profiles show `copyFromSliceVector` as the dominant source of memory consumption.

**Comment by rytaft** [comment]
It seems to me that `DecodeTableValue` may keep some references to the underlying byte array for string types.

**Comment by rytaft** [comment]
Actually, `DecodeTableKey` may also keep some references, as described [here](https://github.com/cockroachdb/cockroach/blob/35e138aa3c2be545fb4e17a85ea6f1b8d6525e53/pkg/util/encoding/encoding.go#L754). 
> Note that the returned string may share storage with the input buffer.

**Comment by rytaft** [comment]
I opened #42357, which is a low-risk change to sample only the columns needed for histogram creation. The problem that some decoded datums may reference the encoded bytes still remains, though it should be mitigated by this change. 

It's very difficult to know if a particular datum still references the underlying array of 10000 rows, so it seems to me that the only way to completely fix the issue (without drastically changing the kv interface or the way we do stats collection) is to perform a deep copy of every datum of type string, []byte, etc inside the `SampleReservoir`. This seems risky, since it could have performance implications, and it's not clear to me how often datums actually reference the byte array anyway.

I welcome any ideas/suggestions...

**Comment by RaduBerinde** [comment]
I'm not too worried about the performance impact of such a change. The chance of a row being selected for the reservoir gets smaller and smaller as the scan proceeds. Eg after 100k rows, only 1 in 10 will incur that overhead; after 1M rows, only 1 in 100 and so on.

> It seems to me that DecodeTableValue may keep some references to the underlying byte array for string types.

Can you point this out? I see it doing `a.NewDString(tree.DString(data))` which does make a copy.

> Actually, DecodeTableKey may also keep some references, as described here.

Nice find! I only looked at the values path, but it is possible that the PK has a string column.

**Comment by rytaft** [comment]
> I'm not too worried about the performance impact of such a change.

Ok, makes sense. I'll make that change in an additional commit on the same PR.

> Can you point this out? I see it doing a.NewDString(tree.DString(data)) which does make a copy.

[Here](https://github.com/cockroachdb/cockroach/blob/35e138aa3c2be545fb4e17a85ea6f1b8d6525e53/pkg/sql/sqlbase/datum_alloc.go#L88) is the code for `NewDString`. Looks like it's not doing a deep copy of the string.

I think it would need to do something like this: 
```
	c := make([]byte, len(v))
	copy(c, v)
	*r = string(c)
```

**Comment by RaduBerinde** [comment]
`tree.DString(data)` does make a copy (`string`s are immutable and `data` is a `[]byte`).

**Comment by tim-o** [comment]
> > I would suggest setting `sql.stats.automatic_collection.enabled` to false to see if that stabilizes the cluster.
> 
> I've made this change as of this comment.. Will check how things look on Monday.

I've heard through the grapevine that disabling stats collection stabilized the cluster.

**Comment by tim-o** [comment]
Do we need to re-enable it and collect a profile or is that enough to confirm what's wrong?

**Comment by RaduBerinde** [comment]
I think we have a profile already. I would disable histograms and reenable stat collection on the cluster going forward.

**Comment by tim-o** [comment]
What's the cluster setting for histograms? I don't see it under `sql.stats...` [here](https://www.cockroachlabs.com/docs/stable/cluster-settings.html).

**Comment by RaduBerinde** [comment]
`sql.stats.histogram_collection.enabled` (you can see it if you select the 19.2 dev version)

**Comment by rytaft** [comment]
Great point! I've looked through the code and I can't find any other types besides `DString` (possibly wrapped in an `OidWrapper`?) and `DCollatedString` that might be referencing the kv batch. And it seems the problem only occurs if the column is key encoded.

**Comment by rytaft** [comment]
The fixes are all merged to master and backported to the 19.2 branch. They will be available in the first patch release of 19.2 (or we can make a custom build earlier if that seems useful). Either way, it would be helpful to try re-enabling histogram collection at that point to see if the problem is fixed for this cluster.

**Comment by ricardocrdb** [comment]
@naveenwashere Since the ZenDesk ticket is closed, I will go ahead and close this issue as well. Please see the below update and PR# [42380](https://github.com/cockroachdb/cockroach/pull/42380) for details of the fix.

> The fixes are all merged to master and backported to the 19.2 branch. They will be available in the first patch release of 19.2 (or we can make a custom build earlier if that seems useful). Either way, it would be helpful to try re-enabling histogram collection at that point to see if the problem is fixed for this cluster.
